# Step one: Cleaning

In this notebook we will focus on taking the original data file and reducing all the errors in it. This will produce a clean file that will allow for meaningful analysis.

## Importing libraries and reading the file

In [117]:
import pandas as pd

attacks = pd.read_csv("data/attacks.csv", encoding='unicode_escape')

#We will alos run the 'cleaning.py' script as it will define functions we will use

%run -i 'src/cleaning.py'

## Initial exploration
Now we will start exploring the file. Using the ".head()" we can get a preview of the columns and a the first rows of the dataset.

In [118]:
attacks.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                                                     href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

                                                                             href  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

  Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

In [119]:
attacks.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [120]:
attacks.shape

(25723, 24)

Using 'head()' , '.columns' and '.shape' we can see that we are dealing with a file of 24 columns (names listed above) and 25723 entries. Lets explore further:

In [121]:
attacks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

From this output we gather a few things:

#### Types and names of the columns: 
* There is only two columns with floats
    * This is odd considering that the column "Age" should also be a number
    * The column "Original order" probably adds little to no value
* Columns 0, 19 and 20 all reference a case number
* Columns 17 and 18 could also be duplicated, since they both reference "href"

####  Non-null values:
* Column 0 "Case number" has significantly more non-null values than the rest of the columns
* Columns 22 and 23 are unnamed and have 2 or less non-null values, these columns add no value to the analysis
* Several columns have a non-null count nearing 6300. But most columns have less than this ammount. This means that we are dealing with a great amount of incomplete rows.


### Deleting Unnamed columns:
We will start cleaning the file by deleting the last 2 columns. For this we will use a function declared in the 'cleaning.py' ran in the beginning of this notebook:

In [122]:
delete_columns(attacks,['Unnamed: 22','Unnamed: 23'])

Deleted columns:  ['Unnamed: 22', 'Unnamed: 23']


In [123]:
attacks.sample(5)

Case Number         Date    Year        Type        Country  \
23126           NaN          NaN     NaN         NaN            NaN   
2131   1999.03.14.b  14-Mar-1999  1999.0  Unprovoked  NEW CALEDONIA   
19126           NaN          NaN     NaN         NaN            NaN   
21470           NaN          NaN     NaN         NaN            NaN   
8540              0          NaN     NaN         NaN            NaN   

                  Area Location      Activity             Name Sex   ...  \
23126              NaN      NaN           NaN              NaN  NaN  ...   
2131   Loyalty Islands    Ouvea  Spearfishing  Blaise Wouanena    M  ...   
19126              NaN      NaN           NaN              NaN  NaN  ...   
21470              NaN      NaN           NaN              NaN  NaN  ...   
8540               NaN      NaN           NaN              NaN  NaN  ...   

      Fatal (Y/N)     Time             Species   \
23126         NaN      NaN                  NaN   
2131            N  Morning  200 to 300 kg shark   
19126         NaN      NaN                  NaN   
21470         NaN      NaN                  NaN   
8540          NaN      NaN                  NaN   

                                   Investigator or Source  \
23126                                                 NaN   
2131   W. Leander; Les Nouvelles Caledoniennes, 3/15/1999   
19126                                                 NaN   
21470                                                 NaN   
8540                                                  NaN   

                             pdf  \
23126                        NaN   
2131   1999.03.14.b-Wouanena.pdf   
19126                        NaN   
21470                        NaN   
8540                         NaN   

                                                                          href formula  \
23126                                                                              NaN   
2131   http://sharkattackfile.net/spreadsheets/pdf_directory/1999.03.14.b-Wouanena.pdf   
19126                                                                              NaN   
21470                                                                              NaN   
8540                                                                               NaN   

                                                                                  href  \
23126                                                                              NaN   
2131   http://sharkattackfile.net/spreadsheets/pdf_directory/1999.03.14.b-Wouanena.pdf   
19126                                                                              NaN   
21470                                                                              NaN   
8540                                                                               NaN   

      Case Number.1 Case Number.2 original order  
23126           NaN           NaN            NaN  
2131   1999.03.14.b  1999.03.14.b         4172.0  
19126           NaN           NaN            NaN  
21470           NaN           NaN            NaN  
8540            NaN           NaN            NaN  

[5 rows x 22 columns]

### Dealing with the "Case Number" columns
Our data set contains 3 columns referring to a case number. For this we will use a function declared in the 'cleaning.py' ran in the beginning of this notebook: 

In [124]:
compare_columns(attacks,["Case Number","Case Number.1","Case Number.2"])

Column information:


Columns            Case Number  Case Number.1  Case Number.2
Lengths                  25723          25723          25723
Non - null values         8702           6302           6302

Data type comparison:


Case Number  Case Number.1  Case Number.2
<class 'float'>        17021          19421          19421
<class 'str'>           8702           6302           6302

Value comparisons:


Case Number  Case Number.1  Case Number.2
Case Number          25723          23298          23318
Case Number.1        23298          25723          25703
Case Number.2        23318          25703          25723

The above matrix shows the amount of rows that each column shares with other columns. E.g. how many cells store the same information in both columns
Sample of columns with differences:


Case Number Case Number.1 Case Number.2
21903         NaN           NaN           NaN
16865         NaN           NaN           NaN
13259         NaN           NaN           NaN
14311         NaN           NaN           NaN
24699         NaN           NaN           NaN

We can see that the columns "Case Number.1" and "Case Number.2" hold almost the same information (They return the same values in all the tests, and in the comparison matrix they only have 20 values that are different). Lets check these values:

In [125]:
different_value_columns(attacks, ["Case Number.1", "Case Number.2"])

Columns differ in the following 20 values


Case Number.1 Case Number.2
34      2018.04.02    2018.04.03
117   2017/07.20.a  2017.07.20.a
144     2017.06.06    2017.05.06
217     2016.09.16    2016.09.15
314   2015.01.24.b  2016.01.24.b
334     2015.11.07    2015.12.23
339     2015.10.28  2015.10.28.a
560     2013.05.04    2014.05.04
3522    1967/07.05    1967.07.05
3795  1962.08.30.b  1962,08.30.b
3880  1961.09,06.R  1961.09.02.R
4403    1952.08.04    1952.08.05
4425  1951.12.15.R  1851.12.15.R
5275  1919.00.00.R  1900.00.00.R
5358  1913.17.10.R  1913.07.10.R
5386  1911.07.31.T  1911.07.31.R
5777  1884.04.28.R  1884.04.38.R
5883  1874.06.18.R  1876.06.18.R
6009    1853.04.29    1853,94.29
6122  1808.05.01.R  1808.06.01.R

We can see that the columns hold mostly the same information. Most of the differences are symbols (commas for periods or slashes) and missing letters. We will keep column Case Number.2 and drop Case Number.1

In [126]:
attacks.drop(["Case Number.1"], axis=1, inplace=True)
attacks.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

                                                      Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and paddle damaged           N   
1                                 Minor injury to left thigh           N   
2               Injury to left lower leg from surfboard skeg           N   
3                                  Minor injury to lower leg           N   
4          Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                                                     href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

                                                                             href  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

  Case Number.2 original order  
0    2018.06.25         6303.0  
1    2018.06.18         6302.0  
2    2018.06.09         6301.0  
3    2018.06.08         6300.0  
4    2018.06.04         6299.0  

[5 rows x 21 columns]

Now lets compare "Case Number" with "Case Number.2"

In [127]:
different_value_columns(attacks, ["Case Number.2", "Case Number"])

Columns differ in the following 2405 values


Case Number.2   Case Number
390      2015.07.10    2015.07-10
4949   1934.02.08.R  1934.01.08.R
5488   1905.09.06.R              
5944     1864.05.00       1864.05
6302                            0
...             ...           ...
8698                            0
8699                            0
8700                            0
8701                            0
25722                          xx

[2405 rows x 2 columns]

We can see that "Case Number" has a great deal of 0 values, lets print the value_counts for "Case Number" where it is different from "Case Number.2"

In [128]:
print_test = different_value_columns(attacks, ["Case Number.2", "Case Number"])
print_test["Case Number"].value_counts()

Columns differ in the following 2405 values


0               2400
2015.07-10         1
1934.01.08.R       1
                   1
1864.05            1
xx                 1
Name: Case Number, dtype: int64

As we suspected, "Case Number" has 2400 values where it holds "0", one "xx" value and only 4 cases where it is similar to "Case Number.2". Since "Case Number.2" holds more complete information in these values and "Case Number" adds no value with these 0 values, we will drop it.

In [129]:
attacks.drop(["Case Number"], axis=1, inplace=True)
attacks.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                                      Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and paddle damaged           N   
1                                 Minor injury to left thigh           N   
2               Injury to left lower leg from surfboard skeg           N   
3                                  Minor injury to lower leg           N   
4          Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                                                     href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

                                                                             href  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

  Case Number.2  original order  
0    2018.06.25          6303.0  
1    2018.06.18          6302.0  
2    2018.06.09          6301.0  
3    2018.06.08          6300.0  
4    2018.06.04          6299.0

Now lets explore the columsn "href formula" and "href" since they appear to hold the same information. Since they hold web addresses, I'll set the column width to be greater in order to be able to see the full information.

In [130]:
pd.set_option("max_colwidth", 500)
different_value_columns(attacks, ["href formula", "href"])

Columns differ in the following 60 values


href formula  \
50                      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.01.13-Stewart.pdf   
96                     http://sharkattackfile.net/spreadsheets/pdf_directory/2017.08.27-Brundler.pdf   
131             http://sharkattackfile.net/spreadsheets/pdf_directory/2017.06.05-FrenchPolynesia.pdf   
133                        http://sharkattackfile.net/spreadsheets/pdf_directory/2017.06.11-Goff.pdf   
141                     http://sharkattackfile.net/spreadsheets/pdf_directory/2017.05.27-Selwood.pdf   
168                   http://sharkattackfile.net/spreadsheets/pdf_directory/2017.03.19-Fernandez.pdf   
234                       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.07.29-Spain.pdf   
241                  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.07.23.a-Cutbirth.pdf   
276                      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.05.21.a-Girl.pdf   
324                    http://sharkattackfile.net/spreadsheets/pdf_directory/2015.12.21.a-Brazil.pdf   
332                  http://sharkattackfile.net/spreadsheets/pdf_directory/2015.11.15.a-Engelman.pdf   
352                 http://sharkattackfile.net/spreadsheets/pdf_directory/2015.10.05.a-Slaughter.pdf   
401                 http://sharkattackfile.net/spreadsheets/pdf_directory/2015.06.27.a-Swanepoel.pdf   
453                      http://sharkattackfile.net/spreadsheets/pdf_directory/2015.01.24-Murray.pdf   
484                        http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.19-Tara.pdf   
485                    http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.18-Roberson.pdf   
486                     http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.17-Olsson, pdf   
493                   http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.03.b-Kayaks2.pdf   
494                   http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.03.a-Kayaks1.pdf   
515                       http://sharkattackfile.net/spreadsheets/pdf_directory/2014.08.15-Hayes.pdf   
527                       http://sharkattackfile.net/spreadsheets/pdf_directory/2014.07.22-Rintz.pdf   
545                     http://sharkattackfile.net/spreadsheets/pdf_directory/2014.06.09.a-Berry.pdf   
569                   http://sharkattackfile.net/spreadsheets/pdf_directory/2014.04.03-Armstrong.pdf   
583                    http://sharkattackfile.net/spreadsheets/pdf_directory/2014.02.07.b-Torres.pdf   
584                 http://sharkattackfile.net/spreadsheets/pdf_directory/2014.02.07-PorpoiseBay.pdf   
585                     http://sharkattackfile.net/spreadsheets/pdf_directory/2014.01.26-O'Connor.pf   
586                           http://sharkattackfile.net/spreadsheets/pdf_directory/2014.01.25-Grant   
587                      http://sharkattackfile.net/spreadsheets/pdf_directory/2014.01.04-Rianto.pdf   
588           http://sharkattackfile.net/spreadsheets/pdf_directory/2014.00.00.b-OceanicWhitetip.pdf   
590                http://sharkattackfile.net/spreadsheets/pdf_directory/2013.12.25-NewCaledonia.pdf   
600                    http://sharkattackfile.net/spreadsheets/pdf_directory/2013.11.22-Gardiner.pdf   
601                        http://sharkattackfile.net/spreadsheets/pdf_directory/2013.11.12-Daly.pdf   
602                     http://sharkattackfile.net/spreadsheets/pdf_directory/2013.11.10-Bahamas.pdf   
603                     http://sharkattackfile.net/spreadsheets/pdf_directory/2013.11.07.b-Gates.pdf   
604                    http://sharkattackfile.net/spreadsheets/pdf_directory/2013.11.07.a-Melian.pdf   
605            http://sharkattackfile.net/spreadsheets/pdf_directory/2013.10.31-ChristianBrandon.pdf   
606                     http://sharkattackfile.net/spreadsheets/pdf_directory/2013.10.28-Simmons.pdf   
607                    http://sharkattackfile.net/spreadsheets/pdf_directory/2013.10.26.b-Tanguy.pdf   
671                             

We can see that most of the differences are due to differences in the case number. We can also see that href have several instances where it repeats the web address. Since this column adds little value to the analysis we will conduct later, for simplicity I will fill the missing values of "href formula" with the values in "href" and drop the "href" column.

In [131]:
attacks.iloc[3245,16] = attacks.iloc[3245,17]
attacks.iloc[3244,16] = attacks.iloc[3244,17]
attacks.drop(["href"], axis=1, inplace=True)
attacks.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                                      Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and paddle damaged           N   
1                                 Minor injury to left thigh           N   
2               Injury to left lower leg from surfboard skeg           N   
3                                  Minor injury to lower leg           N   
4          Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                                                     href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

  Case Number.2  original order  
0    2018.06.25          6303.0  
1    2018.06.18          6302.0  
2    2018.06.09          6301.0  
3    2018.06.08          6300.0  
4    2018.06.04          6299.0

Now that we have altered our data frame a bit, lets tidy up. Ill reorder and rename a few columns to maintain structre and then we will check the shape and non-null count again.

In [140]:
attacks = attacks[['Case Number.2', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location', 'Activity',
       'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time', 'Species ',
       'Investigator or Source', 'pdf', 'href formula',
       'original order']]
attacks.rename(columns={'href formula' : 'href',
                        'Case Number.2' : 'Case Number'}, inplace=True, errors='raise')
attacks.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                                      Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and paddle damaged           N   
1                                 Minor injury to left thigh           N   
2               Injury to left lower leg from surfboard skeg           N   
3                                  Minor injury to lower leg           N   
4          Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                                                             href  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

   original order  
0          6303.0  
1          6302.0  
2          6301.0  
3          6300.0  
4          6299.0